lark/parser_frontends.py

```diff
@@ -93,26 +105,87 @@ def _verify_start(self, start=None):
         chosen_start = self._verify_start(start)
         kw = {} if on_error is None else {'on_error': on_error}
-        stream = self._make_lexer_thread(text)
+        stream = self._make_lexer_thread(text, start_pos=start_pos, end_pos=end_pos)
         return self.parser.parse(stream, chosen_start, **kw)
 
-    def parse_interactive(self, text: Optional[str]=None, start=None):
+    def parse_interactive(self, text: Optional[str]=None, start=None,
+                          *, start_pos: Optional[int] = None, end_pos: Optional[int] = None):
         # TODO BREAK - Change text from Optional[str] to text: str = ''.
         #   Would break behavior of exhaust_lexer(), which currently raises TypeError, and after the change would just return []
+        #   When this is done, also adjust the code in `LexerState.__init__` since it currently works around being
+        #   passed `None` with regard to start_pos and end_pos
         chosen_start = self._verify_start(start)
         if self.parser_conf.parser_type != 'lalr':
             raise ConfigurationError("parse_interactive() currently only works with parser='lalr' ")
-        stream = self._make_lexer_thread(text)  # type: ignore[arg-type]
+        stream = self._make_lexer_thread(text, start_pos=start_pos, end_pos=end_pos)  # type: ignore[arg-type]
         return self.parser.parse_interactive(stream, chosen_start)
 
 
+    def scan(self, text: str, start: Optional[str]=None, *, start_pos: Optional[int] = None,
+             end_pos: Optional[int] = None) -> Iterable[ScanMatch]:
+        """
+        In contrast to the other functions here, this one actually does work. See `Lark.scan`
+        for a description of what this function is for.
+        """
+        if self.options.parser != 'lalr':
+            raise ValueError("scan requires parser='lalr'")
+        start_states = self.parser._parse_table.start_states
+        chosen_start = self._verify_start(start)
+        start_state = start_states[chosen_start]
+        pos = start_pos if start_pos is not None else 0
+        end_pos = end_pos if end_pos is not None else len(text)
+        if pos < 0:
+            pos += len(text)
+        if end_pos < 0:
+            end_pos += len(text)
+        while True:
+            # Find the next candidate location
+            found = self.lexer.search_start(text, start_state, pos, end_pos)
+            # No more valid candidates
+            if found is None:
+                break
+            assert found.end_pos <= end_pos
+            # Collect the potential end points found for this parse
+            # We need to keep track of multiple options in case there are false `$END`s in the `ip.choices()`
+            # We don't want to check early since this can be expensive.
+            valid_end = []
+            ip = self.parse_interactive(text, start=chosen_start, start_pos=found.start_pos, end_pos=end_pos)
+            tokens = ip.lexer_thread.lex(ip.parser_state)
```

Comment:
Why not use `ip.iter_parse()`?

Reply:
I had in mind that it doesn't work, gotta check.